File: Jan-2023/Java/SameTree.java
100. Same Tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }f
 * }
 */
class Solution {
	public boolean isSameTree(TreeNode p, TreeNode q) {
		//both TreeNodes are null
		if (p == null && q == null) return true;
		//any one TreeNode is null
		else if (p == null || q == null) return false;
		//left and right TreeNode are not same in value
		else if (p.val != q.val) return false;
		else {
			//check for left and right branch
			return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
		}
	}
}

File: Jan-2023/C++/SameTree.cxx
// 100. Same Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
  public:
	bool isSameTree(TreeNode *p, TreeNode *q)
	{
		//both treenode are null
		if (p == nullptr && q == nullptr)
		{
			return true;
		}
		//anyone of them is null, so it means they are not the same
		if (p == nullptr || q == nullptr)
		{
			return false;
		}
		//both of them have different values
		else if (p->val != q->val)
		{
			return false;
		}
       //check left branch and also another revursive call for right branch
		return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
	}
};
